In [1]:
import pandas as pd

df = pd.read_csv("../data/raw/livingcost_india_all_inr (4).csv")
df1 = pd.read_csv("../data/raw/india_cost_quality_dataset.csv")
df.head()

,City,cost_one_person_usd,rent_one_person_usd,monthly_salary_after_tax_usd,income_after_rent_usd,months_covered,cost_one_person_inr,rent_one_person_inr,monthly_salary_after_tax_inr,income_after_rent_inr,usd_to_inr_rate_used,source_url
0,Adoni,346,132.0,329,197.0,0.9,30400.87,11598.02,28907.19,17309.17,87.8638,https://livingcost.org/cost/india/ap/adoni
1,Anantapur,345,136.0,571,435.0,1.7,30313.01,11949.48,50170.23,38220.75,87.8638,https://livingcost.org/cost/india/ap/anantapur
2,Chirala,295,102.0,555,453.0,1.8,25919.82,8962.11,48764.41,39802.30,87.8638,https://livingcost.org/cost/india/ap/chirala
3,Chittoor,400,81.5,404,322.5,1.4,35145.52,7160.90,35496.98,28336.08,87.8638,https://livingcost.org/cost/india/ap/chittoor
4,Eluru,298,89.2,384,294.8,1.3,26183.41,7837.45,33739.70,25902.25,87.8638,https://livingcost.org/cost/india/ap/eluru


In [ ]:
df.shape

(221, 12)

In [2]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 221 entries, 0 to 220
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   City                          221 non-null    str    
 1   cost_one_person_usd           221 non-null    int64  
 2   rent_one_person_usd           221 non-null    float64
 3   monthly_salary_after_tax_usd  221 non-null    int64  
 4   income_after_rent_usd         221 non-null    float64
 5   months_covered                221 non-null    float64
 6   cost_one_person_inr           221 non-null    float64
 7   rent_one_person_inr           221 non-null    float64
 8   monthly_salary_after_tax_inr  221 non-null    float64
 9   income_after_rent_inr         221 non-null    float64
 10  usd_to_inr_rate_used          221 non-null    float64
 11  source_url                    221 non-null    str    
dtypes: float64(8), int64(2), str(2)
memory usage: 20.8 KB


In [ ]:
df.columns.to_list()

['City',
 'cost_one_person_usd',
 'rent_one_person_usd',
 'monthly_salary_after_tax_usd',
 'income_after_rent_usd',
 'months_covered',
 'cost_one_person_inr',
 'rent_one_person_inr',
 'monthly_salary_after_tax_inr',
 'income_after_rent_inr',
 'usd_to_inr_rate_used',
 'source_url']

In [ ]:
df.shape

In [ ]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 221 entries, 0 to 220
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   City                          221 non-null    str    
 1   cost_one_person_usd           221 non-null    int64  
 2   rent_one_person_usd           221 non-null    float64
 3   monthly_salary_after_tax_usd  221 non-null    int64  
 4   income_after_rent_usd         221 non-null    float64
 5   months_covered                221 non-null    float64
 6   cost_one_person_inr           221 non-null    float64
 7   rent_one_person_inr           221 non-null    float64
 8   monthly_salary_after_tax_inr  221 non-null    float64
 9   income_after_rent_inr         221 non-null    float64
 10  usd_to_inr_rate_used          221 non-null    float64
 11  source_url                    221 non-null    str    
dtypes: float64(8), int64(2), str(2)
memory usage: 20.8 KB


In [ ]:
df.describe()

,cost_one_person_usd,rent_one_person_usd,monthly_salary_after_tax_usd,income_after_rent_usd,months_covered,cost_one_person_inr,rent_one_person_inr,monthly_salary_after_tax_inr,income_after_rent_inr,usd_to_inr_rate_used
count,221.000000,221.000000,221.000000,221.000000,221.000000,221.000000,221.000000,221.000000,221.000000,2.210000e+02
mean,357.932127,119.447059,357.796380,238.349321,1.095475,31449.276197,10495.072342,31437.349338,20942.277133,8.786380e+01
std,91.054379,47.159346,104.077197,114.603182,0.360653,8000.383058,4143.599201,9144.618287,10069.470970,1.424312e-14
min,251.000000,59.700000,179.000000,-187.000000,0.300000,22053.810000,5245.470000,15727.620000,-16430.530000,8.786380e+01
25%,301.000000,92.800000,280.000000,170.000000,0.900000,26447.000000,8153.760000,24601.860000,14936.850000,8.786380e+01
50%,329.000000,112.000000,360.000000,248.000000,1.100000,28907.190000,9840.750000,31630.970000,21790.220000,8.786380e+01
75%,382.000000,129.000000,406.000000,301.000000,1.300000,33563.970000,11334.430000,35672.700000,26447.000000,8.786380e+01
max,705.000000,458.000000,712.000000,621.700000,2.700000,61943.980000,40241.620000,62559.030000,54624.920000,8.786380e+01


In [ ]:
df.isnull().sum()

City                            0
cost_one_person_usd             0
rent_one_person_usd             0
monthly_salary_after_tax_usd    0
income_after_rent_usd           0
months_covered                  0
cost_one_person_inr             0
rent_one_person_inr             0
monthly_salary_after_tax_inr    0
income_after_rent_inr           0
usd_to_inr_rate_used            0
source_url                      0
dtype: int64

In [ ]:
df.duplicated().sum()

np.int64(0)

In [ ]:
df.nunique()

City                            221
cost_one_person_usd             100
rent_one_person_usd             134
monthly_salary_after_tax_usd    166
income_after_rent_usd           198
months_covered                   19
cost_one_person_inr             100
rent_one_person_inr             134
monthly_salary_after_tax_inr    167
income_after_rent_inr           198
usd_to_inr_rate_used              1
source_url                      221
dtype: int64

In [ ]:
df.select_dtypes(include="number").corr()

,cost_one_person_usd,rent_one_person_usd,monthly_salary_after_tax_usd,income_after_rent_usd,months_covered,cost_one_person_inr,rent_one_person_inr,monthly_salary_after_tax_inr,income_after_rent_inr,usd_to_inr_rate_used
cost_one_person_usd,1.000000,0.266667,-0.179571,-0.272812,-0.298656,1.000000,0.266667,-0.179571,-0.272812,NaN
rent_one_person_usd,0.266667,1.000000,-0.007928,-0.418701,-0.402955,0.266667,1.000000,-0.007928,-0.418701,NaN
monthly_salary_after_tax_usd,-0.179571,-0.007928,1.000000,0.911415,0.858648,-0.179571,-0.007928,1.000000,0.911415,NaN
income_after_rent_usd,-0.272812,-0.418701,0.911415,1.000000,0.945600,-0.272812,-0.418701,0.911415,1.000000,NaN
months_covered,-0.298656,-0.402955,0.858648,0.945600,1.000000,-0.298657,-0.402955,0.858648,0.945600,NaN
cost_one_person_inr,1.000000,0.266667,-0.179571,-0.272812,-0.298657,1.000000,0.266667,-0.179571,-0.272812,NaN
rent_one_person_inr,0.266667,1.000000,-0.007928,-0.418701,-0.402955,0.266667,1.000000,-0.007928,-0.418701,NaN
monthly_salary_after_tax_inr,-0.179571,-0.007928,1.000000,0.911415,0.858648,-0.179571,-0.007928,1.000000,0.911415,NaN
income_after_rent_inr,-0.272812,-0.418701,0.911415,1.000000,0.945600,-0.272812,-0.418701,0.911415,1.000000,NaN
usd_to_inr_rate_used,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
df["City"].head(20).tolist()

[' Adoni',
 ' Anantapur',
 ' Chirala',
 ' Chittoor',
 ' Eluru',
 ' Guntur',
 ' Hindupur',
 ' Kadapa',
 ' Kakinada',
 ' Kurnool',
 ' Machilipatnam',
 ' Madanapalle',
 ' Nandyal',
 ' Nellore',
 ' Ongole',
 ' Proddatur',
 ' Rajahmundry',
 ' Tenali',
 ' Tirupati',
 ' Vijayawada']

In [ ]:
df["City"].tail(20).tolist()

[' Kolkata',
 ' Krishnanagar',
 ' Kulti',
 ' Madhyamgram',
 ' Maheshtala',
 ' Malda',
 ' Midnapore',
 ' Naihati',
 ' North Dumdum',
 ' Panihati',
 ' Raiganj',
 ' Rajarhat',
 ' Rajpur Sonarpur',
 ' Serampore',
 ' Shantipur',
 ' Siliguri',
 ' South Dumdum',
 ' Uluberia',
 ' Uttarpara',
 ' Alleppey']

In [ ]:
df["City"] = df["City"].str.strip()
df1["City"] = df1["City"].str.strip()

In [ ]:
df["City"].head(20).tolist()

['Adoni',
 'Anantapur',
 'Chirala',
 'Chittoor',
 'Eluru',
 'Guntur',
 'Hindupur',
 'Kadapa',
 'Kakinada',
 'Kurnool',
 'Machilipatnam',
 'Madanapalle',
 'Nandyal',
 'Nellore',
 'Ongole',
 'Proddatur',
 'Rajahmundry',
 'Tenali',
 'Tirupati',
 'Vijayawada']

In [ ]:
df1["City"].head(20).tolist()

['Mumbai',
 'Delhi',
 'Bengaluru',
 'Hyderabad',
 'Ahmedabad',
 'Chennai',
 'Kolkata',
 'Pune',
 'Jaipur',
 'Surat',
 'Lucknow',
 'Kanpur',
 'Nagpur',
 'Indore',
 'Thane',
 'Bhopal',
 'Visakhapatnam',
 'Pimpri-Chinchwad',
 'Patna',
 'Vadodara']

In [ ]:
common_cities = set(df["City"]) & set(df1["City"])

len(common_cities)

108

In [ ]:
sorted(common_cities)

['Agartala',
 'Agra',
 'Ahmedabad',
 'Aizawl',
 'Ajmer',
 'Aligarh',
 'Allahabad',
 'Alwar',
 'Amritsar',
 'Anantapur',
 'Asansol',
 'Aurangabad',
 'Balurghat',
 'Barasat',
 'Bardhaman',
 'Bareilly',
 'Bhavnagar',
 'Bhilai',
 'Bhopal',
 'Bhubaneswar',
 'Bikaner',
 'Bilaspur',
 'Chandigarh',
 'Chennai',
 'Chittoor',
 'Cuttack',
 'Dehradun',
 'Delhi',
 'Dhanbad',
 'Dibrugarh',
 'Durgapur',
 'Eluru',
 'Faridabad',
 'Gaya',
 'Ghaziabad',
 'Gulbarga',
 'Guntur',
 'Gurgaon',
 'Guwahati',
 'Gwalior',
 'Hisar',
 'Howrah',
 'Imphal',
 'Indore',
 'Jabalpur',
 'Jaipur',
 'Jalandhar',
 'Jalgaon',
 'Jamshedpur',
 'Jhansi',
 'Jodhpur',
 'Kakinada',
 'Kanpur',
 'Kharagpur',
 'Kolhapur',
 'Kolkata',
 'Kollam',
 'Krishnanagar',
 'Kurnool',
 'Lucknow',
 'Ludhiana',
 'Madurai',
 'Malda',
 'Mathura',
 'Meerut',
 'Midnapore',
 'Moradabad',
 'Mumbai',
 'Muzaffarpur',
 'Nagpur',
 'Nanded',
 'Nashik',
 'Navi Mumbai',
 'Nellore',
 'Nizamabad',
 'Pathankot',
 'Patiala',
 'Patna',
 'Pondicherry',
 'Porbandar',
 

In [ ]:
df1.columns.tolist()

['City',
 'Average Rent (INR/month)',
 'Food Cost (INR/month)',
 'Internet Speed (Mbps)',
 'Healthcare Rating',
 'Safety Score',
 'Happiness Index']

In [ ]:
df.columns.tolist()

['City',
 'cost_one_person_usd',
 'rent_one_person_usd',
 'monthly_salary_after_tax_usd',
 'income_after_rent_usd',
 'months_covered',
 'cost_one_person_inr',
 'rent_one_person_inr',
 'monthly_salary_after_tax_inr',
 'income_after_rent_inr',
 'usd_to_inr_rate_used',
 'source_url']

In [ ]:
df1["Average Rent (INR/month)"].describe()

count      170.000000
mean      8022.052941
std       4839.775925
min       4009.000000
25%       5490.750000
50%       6966.000000
75%       8599.750000
max      34896.000000
Name: Average Rent (INR/month), dtype: float64

In [ ]:
df["rent_one_person_inr"].describe()

count      221.000000
mean     10495.072342
std       4143.599201
min       5245.470000
25%       8153.760000
50%       9840.750000
75%      11334.430000
max      40241.620000
Name: rent_one_person_inr, dtype: float64

In [ ]:
# Compare rent for common cities
rent_comparison = df1[df1["City"].isin(common_cities)][
    ["City", "Average Rent (INR/month)"]
].merge(
    df[["City", "rent_one_person_inr"]],
    on="City"
)

rent_comparison.head(20)

,City,Average Rent (INR/month),rent_one_person_inr
0,Mumbai,34896,35233.38
1,Delhi,26424,16694.12
2,Ahmedabad,20881,13970.34
3,Chennai,21323,13970.34
4,Kolkata,25710,10631.52
5,Pune,33540,18363.53
6,Jaipur,8774,10982.98
7,Surat,9697,40241.62
8,Lucknow,10047,9928.61
9,Kanpur,4519,8751.23


In [ ]:
rent_comparison["rent_difference_percent"] = (
    abs(
        rent_comparison["Average Rent (INR/month)"]
        - rent_comparison["rent_one_person_inr"]
    )
    / rent_comparison["rent_one_person_inr"]
) * 100

rent_comparison["rent_difference_percent"].describe()

count    108.000000
mean      34.391124
std       23.862101
min        0.275869
25%       15.368854
50%       34.726882
75%       48.952067
max      141.828074
Name: rent_difference_percent, dtype: float64

In [ ]:
df["rent_one_person_inr"].value_counts().head(55)

rent_one_person_inr
9752.88     6
10016.47    6
10192.20    6
9225.70     5
10631.52    5
11510.16    4
10104.34    4
10982.98    4
14058.21    4
11334.43    4
10455.79    4
11598.02    3
8962.11     3
11246.57    3
9049.97     3
13970.34    3
9577.15     3
11158.70    3
9840.75     3
10895.11    3
10280.06    3
9928.61     3
8874.24     3
12652.39    2
7916.53     2
12125.20    2
8707.30     2
16430.53    2
20472.27    2
8250.41     2
12915.98    2
6387.70     2
9137.84     2
11773.75    2
9489.29     2
8777.59     2
8232.84     2
8136.19     2
7389.35     2
12564.52    2
12213.07    2
13706.75    2
11861.61    2
7881.38     2
10367.93    2
11949.48    1
7160.90     1
7837.45     1
7099.40     1
7415.70     1
26534.87    1
12828.11    1
8144.97     1
9313.56     1
6932.45     1
Name: count, dtype: int64

In [ ]:
df["rent_one_person_inr"].nunique()

134

In [ ]:
df["rent_one_person_inr"].nunique() / len(df)

0.6063348416289592

So only 60.6% of the rent values are unique. That means a lot of different cities have exactly the same rent_one_person_inr.

In [ ]:
df[df["rent_one_person_inr"].duplicated(keep=False)] \
    .sort_values("rent_one_person_inr")[
        ["City", "rent_one_person_inr"]
    ].head(50)

,City,rent_one_person_inr
45,Navsari,6387.70
80,Kollam,6387.70
105,Solapur,7389.35
156,"Salem, Tamil Nadu",7389.35
157,Thanjavur,7881.38
167,Bareilly,7881.38
176,Mirzapur,7916.53
10,Machilipatnam,7916.53
103,Sangli,8136.19
203,Kulti,8136.19


That's not automatically "wrong", but this pattern is suspicious, especially because the values have very precise decimals yet are repeated across unrelated cities. It suggests the column may have been generated/imputed from a limited set of values rather than being independently observed for every city.

In [ ]:
df.groupby("rent_one_person_inr")["City"].agg(list).sort_values(key=lambda x: x.str.len(), ascending=False).head(15)

rent_one_person_inr
10192.20    [Erode, Bally, Baranagar, Barrackpore, Kamarha...
9752.88     [Visakhapatnam, Dispur, Mysore, Jalgaon, Gwali...
10016.47    [Shimoga, Aurangabad, Nashik, Jodhpur, Coimbat...
9225.70     [Rajahmundry, Bhilwara, Pali, Cuddalore, Allah...
10631.52    [Tirupati, Ranchi, Davanagere, Dindigul, Kolkata]
11334.43             [Vadodara, Amritsar, Jhansi, Maheshtala]
10982.98                 [Ongole, Porbandar, Bijapur, Jaipur]
14058.21                   [Panjim, Kalyan, Dehradun, Haldia]
11510.16               [Hindupur, Durg, Patiala, Mahbubnagar]
10104.34    [Madanapalle, Kota, Rajasthan, Madhyamgram, Se...
10455.79    [Ghaziabad, Bidhannagar, Rajpur Sonarpur, Ulub...
11598.02                           [Adoni, Nandyal, Thrissur]
11158.70               [Hubli, Ernakulam, Thiruvananthapuram]
10280.06                    [Cuttack, North Dumdum, Panihati]
10895.11                        [Bhubaneswar, Sikar, Barasat]
Name: City, dtype: object

In [ ]:
df["rent_one_person_inr"].duplicated(keep=False).sum()

np.int64(132)

In [ ]:
df[df["rent_one_person_inr"].duplicated(keep=False)].sort_values("rent_one_person_inr")[["City", "rent_one_person_inr"]].head(30)

,City,rent_one_person_inr
45,Navsari,6387.70
80,Kollam,6387.70
105,Solapur,7389.35
156,"Salem, Tamil Nadu",7389.35
157,Thanjavur,7881.38
167,Bareilly,7881.38
176,Mirzapur,7916.53
10,Machilipatnam,7916.53
103,Sangli,8136.19
203,Kulti,8136.19


So our decision for Dataset 2

Keep:

cost_one_person_inr
rent_one_person_inr
monthly_salary_after_tax_inr
income_after_rent_inr
months_covered

But later, when we build the final model, we'll decide which ones actually belong in the features.
In particular, months_covered and income_after_rent_inr may be derived from the other variables, so we need to be careful about data leakage/redundancy when choosing the ML target.